daemon/policy.go

```diff
@@ -188,6 +188,19 @@ type PolicyAddEvent struct {
 
 // Handle implements pkg/eventqueue/EventHandler interface.
 func (p *PolicyAddEvent) Handle(res chan interface{}) {
+
+	// This check is needed to avoid a race where a DNS update to a rule
+	// enqueues an copy of that rule that is older than a k8s/API update to
+	// that same rule.  Rules are assigned random UUIDs in MarkToFQDNsRule when
+	// initially inserted and DNS updates always replace existing rules. In
+	// this out-of-order scenario the k8s/API updated rule would receive a new
+	// UUID and the older UUID, belonging to the DNS updated rule, would no
+	// longer be present. This check fails in that case, and relies on the
+	// Add/Remove of a toFQDNs rule in RuleGen happening atomically.
+	if !p.d.dnsRuleGen.AreFQDNRulesStillPresent(p.rules) {
```

Comment:
I just realised an issue with this. The DNS updates include all rules that refer to a DNS name. Some may have been replaced, others may still be there. `AreFQDNRulesStillPresent` would return false if any rule has been replaced, but the others still need to go through. This might mean we need to filter out rules that are no longer valid :/